lucet-concurrency-tests/tests/killswitch.rs

```diff
@@ -0,0 +1,731 @@
+            "do_nothing",
+            FunctionPointer::from_usize(do_nothing as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "run_hostcall",
+            FunctionPointer::from_usize(run_hostcall as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "run_yielding_hostcall",
+            FunctionPointer::from_usize(run_yielding_hostcall as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "fatal",
+            FunctionPointer::from_usize(fatal as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "hit_sigstack_guard_page",
+            FunctionPointer::from_usize(hit_sigstack_guard_page as usize),
+        ))
+        .build()
+}
+
+/// This test hostcall just needs to exist so that we can call it. `LockTestpoints` provide points
+/// for us to ensure that tests that need to happen inside a hostcall, will happen inside a
+/// hostcall.
+#[lucet_hostcall]
+#[no_mangle]
+pub fn real_hostcall(_vmctx: &mut Vmctx) -> bool {
+    true
+}
+
+/// This test hostcall will immediately yield. This is used to test termination of a
+/// yielded instance.
+#[lucet_hostcall]
+#[no_mangle]
+pub fn yielding_hostcall(vmctx: &mut Vmctx) {
+    vmctx.yield_();
+}
+
+pub fn test_c_with_instrumented_guest_entry<F, R>(dir: &str, cfile: &str, f: F) -> R
+where
+    F: FnOnce(InstanceHandle) -> R,
+{
+    test_ex(|| {
+        unsafe {
+            ENTERING_GUEST = Some(Syncpoint::new());
+        }
+        let module = test_module_c(dir, cfile).expect("build and load module");
+        let region = MmapRegion::create(1, &Limits::default()).expect("region can be created");
+
+        let inst = region
+            .new_instance(module)
+            .expect("instance can be created");
+
+        f(inst)
+    })
+}
+
+pub fn test_exclusive_instance_with_instrumented_guest_entry<F, R>(f: F) -> R
+where
+    F: FnOnce(InstanceHandle) -> R,
+{
+    test_ex(|| {
+        unsafe {
+            ENTERING_GUEST = Some(Syncpoint::new());
+        }
+        let module = mock_killswitch_module();
+        let region = MmapRegion::create(1, &Limits::default()).expect("region can be created");
+
+        let inst = region
+            .new_instance(module)
+            .expect("instance can be created");
+
+        f(inst)
+    })
+}
+
+pub fn test_instance_with_instrumented_guest_entry<F, R>(f: F) -> R
+where
+    F: FnOnce(InstanceHandle) -> R,
+{
+    test_nonex(|| {
+        let module = mock_killswitch_module();
+        let region = MmapRegion::create(1, &Limits::default()).expect("region can be created");
+
+        let inst = region
+            .new_instance(module)
+            .expect("instance can be created");
+
+        f(inst)
+    })
+}
+
+// Test that a timeout that occurs in a signal handler is handled cleanly without signalling the
+// Lucet embedder.
+#[test]
+fn terminate_in_guest() {
```

Comment:
Does this case map to the docs description of "Terminated while handling guest fault?"

Reply:
The comment here was a lie :grimacing: There isn't a specific "while handling guest fault" test here, yet! That's the next PR as soon as this lands :)